### DISCUSSION TITLE
[RFC] expo-system-ui module

### DISCUSSION THREAD

**Discussion by EvanBacon** [expo/expo#14286] [Proposals and Ideas]
> The Android navigation bar related functionality will be added via `expo-navigation-bar`: https://github.com/expo/expo/pull/14689

> ✅ Completed in https://github.com/expo/expo/pull/14952!

# Why

Android managed EAS apps cannot configure the system navigation bar without installing a third-party package, and the [androidNavigationBar](https://docs.expo.dev/versions/latest/config/app/#androidnavigationbar) property in the app.json doesn't work -- this means `eas build -p android` doesn't have full parity with legacy `expo build:android` apps.

## Background

The "System UI" refers to features that have a physical display, but fall outside of the view-tree / rendering system in React Native.  This includes features like:

| Feature                    | iOS | Android | Required for EAS |
| -------------------------- | --- | ------- |-------|
| Root view background color | ✅  | ✅      | ❌ |
| Status Bar                 | ✅  | ✅      | ❌ |
| Global appearance          | ✅  | ✅      |✅      |
| Global Tint Color          | ✅  | ✅      | ❌ |
| System navigation bar      | ❌  | ✅      |✅      |
| Task Manager               | *✅ | ✅      | ❌ |

- Root view background color: Visible during orientation changes and when all views have transparent background colors.
- Status bar: Currently managed by `react-native` / `expo-status-bar`.
- Global appearance: setting dark mode / light mode for the entire app.
- Global tint color: Used for changing modal text color, and store review star colors.
- System navigation bar: bottom bar with back button, Android only.
- Task Manager: On iOS, can statically be configured using config plugins to change the icon and name, but Android can be dynamically configured (top bar / icon appearance).

# Proposal

We should add a module `expo-system-ui` which surfaces a JS API for [Android system UI](https://developer.android.com/training/system-ui) API which we currently implement in [Expo Go here](https://github.com/expo/expo/blob/master/android/expoview/src/main/java/host/exp/exponent/utils/ExperienceActivityUtils.kt). 

We could potentially scale this API to also support dynamically changing the rootViewBackgroundColor on iOS/Android, locking the system appearance, setting global tintColor, setting the task description, and other system UI settings that are too small and specific to have their own module.

To support the `androidNavigationBar` object in the `app.json`, we can add a config plugin which writes the static values to `strings.xml` properties, the `expo-system-ui` module would then read those values and apply them as the default settings automatically via the ReactActivityLifecycleListener of `expo-modules-core` (like we do for [expo-splash-screen](https://github.com/expo/expo/pull/14061)).

This module would be Android only, no iOS or web component. I've also checked the w3 spec for a similar mechanism in web to base the native API off of, but didn't find anything.

In its current state, we'd need to include `expo-system-ui` by default in every app to ensure that `androidNavigationBar` isn't being written with nothing to read it. This is obviously undesirable as it adds functionality that a portion of users might not use. 

To counter this, I also propose that we drop `androidNavigationBar` in favor of an Expo config plugin, this would enable us to make `expo-system-ui` optional so users would need to install it in order to configure the system UI. 

```js
{
  "plugins": [
    [
      "expo-system-ui", { 
        // Used for both platforms
        backgroundColor: '#4630EB',
        android: {
          // Override the Android property
          backgroundColor: '#FF0000',
        }
      }
    ]
  ]
}
```

Platform specific keys can be used to override common keys:

```js
{
  plugins: [
    [
      "expo-system-ui", { 
        // Used for both platforms
        backgroundColor: '#4630EB',
        android: {
          // Override the Android property
          backgroundColor: '#FF0000',
        }
      }
    ]
  ]
}
```

## Native

On the native side we'd expose two setter methods for Android navigation bar: 
- setNavigationBarColor: sets a value to the Android `activity.window.navigationBarColor`
- setSystemUiVisibility: sets a value to the Android `activity.window.decorView.systemUiVisibility`

And potentially a getter method to read the current settings of the System UI.

For appearance:

- setSystemAppearance

We'll need to extend the delegates to read a static value and set the initial state, like: [splash screen](https://github.com/expo/expo/commit/603216269a06996ba26315f34285d3f43ca22e15). We'll add accompanying config plugins to write these static values.

We'll also need an interface in Expo Go and Expo Dev Client to provide sandboxed settings. For instance, Expo Go would need to provide the specific root view controller to lock the appearance on. 

## Considerations

We'll need to do the same for appearance (ENG-1358), and potentially orientation locking. We may want to cover everything in the [Expo Go module](https://github.com/expo/expo/blob/master/android/expoview/src/main/java/host/exp/exponent/utils/ExperienceActivityUtils.kt) i.e. `setRootViewBackgroundColor`, `setNavigationBar`, `setTaskDescription`, `setHidden`, `setStyle`, `setTranslucent`, `configureStatusBar`, `updateOrientation`, `overrideUiMode`. In which case we may want to use this API in favor of StatusBar, and implement a iOS module.

**Comment by ide** [comment]
> other system UI settings that are too small and specific to have their own module

This is a key point -- when another module makes more sense (like `expo-status-bar`, `expo-splash-screen`, `expo-keyboard`), prefer a dedicated module.

Do you foresee this module being universal or platform-specific? If most of the methods make sense only on one platform, perhaps we want a set of modules like `expo-system-ui` (universal), `expo-android-system-ui`, `expo-ios-system-ui`, and so on.

**Reply by mrousavy** [reply]
Why create separate modules for Android or iOS specific stuff? I'd start with `expo-system-ui` and if there is a functionality only available on one OS I just mark it Android/iOS only, no?

**Reply by ide** [reply]
If a feature is specific to one platform, as opposed to be absent from one platform, it's not a universal feature. An occasional platform-specific method in a module is alright. When we have many of them, a platform-specific module might be a better home. Expo runs on only three platforms right now but I am thinking about a N-platform future.

**Comment by mrbrentkelly** [comment]
One thing to consider as well is that [systemUIVisibility](https://stackoverflow.com/questions/62577645/android-view-view-systemuivisibility-deprecated-what-is-the-replacement) is deprecated in Android 30. There's a new API to control system bars called [WindowInsetsController](https://developer.android.com/reference/android/view/WindowInsetsController) 👍

**Comment by pluscubed** [comment]
Not much to add, other than that [Android edge-to-edge mode](https://developer.android.com/training/gestures/edge-to-edge), i.e. transparent nav bar, should be made possible (like https://github.com/flutter/flutter/pull/81303)

**Reply by mrousavy** [reply]
Hi! I'm working on this here: https://github.com/expo/expo/pull/14468 - does setting the navigation/status bar visibility do what you mean with the "edge-to-edge mode"?


https://user-images.githubusercontent.com/15199031/134015740-b2dd0da5-8fc8-4ff0-9bb9-bdbf658affbe.mov

**Reply by pluscubed** [reply]
Not exactly - "edge-to-edge" means drawing underneath the status/nav bar, like iOS. The status/nav bars are still visible, just overlaid on top.

Expo already draws underneath the status bar by default, so the only change is drawing underneath the navbar.

This requires 2 steps:
1. `WindowCompat.setDecorFitsSystemWindows(window, false)` - [source code here](https://github.com/androidx/androidx/blob/7b9428fbe68a0314c9c460d950aa825ffb5f75f3/core/core/src/main/java/androidx/core/view/WindowCompat.java#L115)
2. Set the navbar color to transparent

So basically Expo should provide a way to invoke the first step.

----

I think the Flutter PR (https://github.com/flutter/flutter/pull/81303) is a good point of reference for what the API can look like. They added support for the following fullscreen modes:

> * **lean back** - a fullscreen mode that brings system overlays into view by tapping anywhere on the screen
> * **immersive** - a fullscreen mode that brings system overlays into view by swiping at the edge of the screen
> * **immersive sticky** - a fullscreen mode that brings system overlays into view by swiping at the edge of the screen
> * **edge to edge** - this makes the application take up the fullscreen, with the overlays hovering on top

For edge-to-edge, it looks something like:
```dart
SystemChrome.setEnabledSystemUIMode(SystemUiMode.edgeToEdge); // corresponds to step 1 above
SystemChrome.setSystemUIOverlayStyle(const SystemUiOverlayStyle(systemNavigationBarColor: Colors.transparent)); // step 2
```

**Reply by mrousavy** [reply]
Thank you @plusCubed for the detailed explanation. Setting navbar to transparent already works, you can pass transparent colors in the set color func. I'll expose the `setDecorFitsSystemWindow` func too under a nice clean API to set the "fullscreen mode" 👍

**Comment by mrousavy** [comment]
PR -> https://github.com/expo/expo/pull/14468

**Comment by yfunk** [comment]
Are there any plans for `expo-navigation-bar` (or a new module) to support hiding the iOS home indicator (`prefersHomeIndicatorAutoHidden`) as mentioned [in the initial PR](https://github.com/expo/expo/pull/14468#issuecomment-922960061)?
